File: leetcode/p00242/p00242.py
class Solution:
    # Time Complexity: O(n)
    # Space Complexity: O(n)
    def isAnagram(self, s: str, t: str) -> bool:
        if len(s) != len(t):
            return False

        memo = {}
        for ch in s:
            if ch not in memo:
                memo[ch] = 1
            else:
                memo[ch] += 1

        for ch in t:
            if ch not in memo:
                return False
            if memo[ch] == 0:
                return False
            memo[ch] -= 1
        
        return True

if __name__ == '__main__':
    s = Solution()

    assert s.isAnagram("anagram", "nagaram") == True
    assert s.isAnagram("fat", "car") == False

File: leetcode/p00242/p00242.go
package p00242

// Time Complexity: O(n)
// Space Complexity: O(n)
func isAnagram(s string, t string) bool {
	if len(s) != len(t) {
		return false
	}

	// rune 可以保存 unicode 編碼
	memo := map[rune]int{}
	for _, ch := range s {
		memo[ch]++
	}

	for _, ch := range t {
		memo[ch]--
		if memo[ch] < 0 {
			return false
		}
	}

	return true
}
